### ISSUE TITLE
Volumes not cleaned up

### ISSUE THREAD

**Issue by vlerenc** [cf-platform-eng/cf-containers-broker#30]
@frodenas Sorry to bother you again. I noticed (in Swarm mode at least; not sure about plain Docker mode) that two volume folders pile up for each launched container that do not get deleted after container removal/termination. Some thousand or so container creations later your persistent disk is full.

The volume folders are located in /var/vcap/store/docker/docker/volumes and /var/vcap/store/cf-containers-broker. I am not even sure why there are two. Maybe because the container is first created without volume and then started in a consecutive call with volume?

INFO -- : Creating Docker container `cf-70a66738-03bc-4208-97a3-cf09fcbac208'...
INFO -- : +-> Create options: {"name"=>"cf-70a66738-03bc-4208-97a3-cf09fcbac208", "Hostname"=>"", "Domainname"=>"", "User"=>"", "AttachStdin"=>false, "AttachStdout"=>true, "AttachStderr"=>true, "Tty"=>false, "OpenStdin"=>false, "StdinOnce"=>false, "Env"=>["MONGODB_USERNAME=2mcvaxrmed915bum", "MONGODB_PASSWORD=qo9wzksiylpmfiqq", "MONGODB_DBNAME=ybf61i3esdr4kayi"], "Cmd"=>["--smallfiles", "--httpinterface"], "Entrypoint"=>nil, "Image"=>"frodenas/mongodb:2.6", "Labels"=>{}, "Volumes"=>{}, "WorkingDir"=>nil, "NetworkDisabled"=>false, "ExposedPorts"=>{}, "HostConfig"=>{"Memory"=>0, "MemorySwap"=>0, "CpuShares"=>nil, "PublishAllPorts"=>false, "Privileged"=>false}}

INFO -- : Starting Docker container `cf-70a66738-03bc-4208-97a3-cf09fcbac208'...
INFO -- : +-> Start options: {"Binds"=>["/var/vcap/store/cf-containers-broker/cf-70a66738-03bc-4208-97a3-cf09fcbac208/data:/data"], "Links"=>[], "LxcConf"=>{}, "Memory"=>0, "MemorySwap"=>0, "CpuShares"=>nil, "PortBindings"=>{"27017/tcp"=>[{"HostPort"=>"37056"}], "28017/tcp"=>[{"HostPort"=>"37057"}]}, "PublishAllPorts"=>false, "Privileged"=>false, "ReadonlyRootfs"=>false, "VolumesFrom"=>[], "CapAdd"=>[], "CapDrop"=>[], "RestartPolicy"=>{"Name"=>"always"}, "Devices"=>[], "Ulimits"=>[]}

This at least fits the observation that the /var/vcap/store/docker/docker/volumes volume was always empty while the /var/vcap/store/cf-containers-broker contained the real data. In any case, even if you delete the container, the volumes get never deleted which becomes a problem after some time. I noticed it just now that we have a couple hundred active containers, but a couple thousand volume corpses lying around.

**Comment by frodenas** [comment]
Good catch. Fixed at https://github.com/cf-platform-eng/cf-containers-broker/commit/cf8bf4734e0dabfe7447997d7af1292e1c3cbc90.

Seems that docker creates a volume directory if the image exposes a volume (like the [mongodb](https://github.com/frodenas/docker-mongodb/blob/master/Dockerfile#L27) one) and you don't explicitly set a volume host bind on create options. Moving the bind to the create options fixes this issue.

I'll cut a new release later today.

**Comment by vlerenc** [comment]
Thank you Ferran.

I applied the change and the folder in /var/vcap/store/docker/docker/volumes gets removed after container termination.

However, the second folder in /var/vcap/store/cf-containers-broker remains on disk and uses considerable space.

**Comment by frodenas** [comment]
@vlerenc What image are you using? On my tests both directories were deleted. Let's see if there's an specific docker image that doesn't shutdown properly.

**Comment by vlerenc** [comment]
I just checked with Redis, i.e. frodenas/redis v2.8.

**Comment by vlerenc** [comment]
One more remark, as always I can only speak for the Swarm setup, i.e. broker is co-deployed with swarm, but the docker nodes run separately on individual machines.

**Comment by frodenas** [comment]
@vlerenc There's one thing that I don't understand. If you are using the swam deployment then why do you need the bound volumes to a host directory? The broker only creates the host directory locally, so at the swam nodes those directories will not exists. Or does docker create them?

**Comment by vlerenc** [comment]
These directories are both created on the Docker nodes and I noticed them only after the disk was almost full. I guess the reason they are there is due to the start options (log in initial description) and maybe this part is not compatible with Swarm and we didn't noticed it so far?

The broker indeed creates empty folders below /var/vcap/store/cf-containers-broker. Naturally (no Docker installation on the broker), the broker has no /var/vcap/store/docker/docker/volumes. So one additional odd thing is that the broker creates empty folders below /var/vcap/store/cf-containers-broker, but the real issue are the not deleted folders below /var/vcap/store/cf-containers-broker on the actual Docker nodes. They consume, e.g. in the case of MongoDB tremendous amounts of disk space.

**Comment by vlerenc** [comment]
@frodenas, may I ask what was the reason for the explicit volume bind to /var/vcap/store/cf-containers-broker in the first place (not leaving it to Docker which seems to effectively also end up on the persistent disk in /var/vcap/store/docker/docker/volumes)?

I ask because I have another question related to volumes that I wanted to raise/if possible discuss with you: How to impose quotas?

Basically, if you have many containers and usage by kind of everybody/too many to talk to about "docketiquette" and no quota imposed, one service instance/user could cannibalize the entire persistent disk space. Quotas are not yet supported by the containers-broker and I understand it's some larger junk of work to do so (I did that myself for our file system service we use), so I wondered whether we could go a shortcut and switch to devicemapper as outlined here http://stackoverflow.com/questions/24391660/limit-disk-size-and-bandwidth-of-a-docker-container.

However, the prerequisite would be to get rid of the externally bound volumes as they cannot be limited at present with Docker, which is why I ask this question in this context (where the externally bound volumes cause the trouble).

I still don't know about the exact consequences of the devicemapper approach, but with externally bound volumes the quota issue couldn't be resolved with the devicemapper approach/proposal anyhow.

In other words, if the explicit volume bind is not necessary, this ticket/issue would resolve automatically, right? And this could open the door to a quota approach via the devicemapper (if I am not mistaken).

**Comment by vlerenc** [comment]
I switched on my local machine to devicemapper (docker -d -s devicemapper --graph /tmp/docker) and created a container. As expected the root FS was limited to 10 GB. The volume however still runs outside and is (with Docker 1.8) located in /tmp/docker/volumes/ which matches my graph option and the observations above (you use a different graph option path in your Docker control script).

That means, simply getting rid of the bind is not automatically resolving the quota question, but I assume it would resolve this issue.

As for the quota, the root FS is located inside the graph option path (in my case /tmp/docker/mnt/d272...59b21/rootfs), so it would land on the persistent disk as well (in your case/with your Docker control script). So, just speculating here, if the Docker images are changed to use the root FS and no explicit volume anymore, they could be set under quota and be persistent.

Not sure though, which further consequences/disadvantages the devicemapper setting has. It is a workaround anyhow because Docker has no nice quota restrictions otherwise.

**Comment by frodenas** [comment]
Are you sure you don't have a `persistent_volumes` property for your catalog plans? This property will instruct the broker to create a host directory and then bind it to the container. This does not make sense for a swarm deployment. So if you don't specify it, then the broker will not create the host directory at `/var/vcap/store/cf-containers-broker`.

Take a look at the examples:
* The single docker broker has a `persistent_volumes` property for the redis container: https://github.com/cf-platform-eng/docker-boshrelease/blob/master/examples/docker-broker-aws.yml#L251
* But the swarm broker doesn't have that property: https://github.com/cf-platform-eng/docker-boshrelease/blob/master/examples/docker-swarm-broker-aws.yml#L259

Can you please check if your catalog has that property?

**Comment by frodenas** [comment]
The reason for the the explicit volume bind is that when I created the docker broker, the docker daemon didn't have a restart policy. That means that if the vm was destroyed, there wasn't any way to recover the data, because the container can not be restarted. Having an external data directory meant that you can create a new container and attach the volume, so there isn't any data loss. Obviously, this has changed since the introduction of the docker restart policies (more on this on my comment at https://blog.starkandwayne.com/2015/09/21/how-to-scale-out-any-cloud-foundry-service/#comment-2266804352).

**Comment by frodenas** [comment]
w/r/t using devicemapper as the storage driver, this can be done right now using the docker bosh release properties, just specify at your deployment manifest:

```yaml
properties:
  docker:
    storage_driver: devicemapper
    storage_options:
      - "dm.basesize=20G"
```
Obviously this will apply to all docker containers, as there is no way actually to specify the size or a particular container.

**Comment by vlerenc** [comment]
Huh, thank you very much. We created the service yamls beginning of this year and I never noticed this when we switched to Swarm. Apologies. I will update them and retest.

**Comment by frodenas** [comment]
Awesome, let me know how it goes please.

**Comment by vlerenc** [comment]
Works, perfect, thank you and sorry again for the confusion around the /var/vcap/store/cf-containers-broker content.
